templates/base/server/.hathora/wrapper.ts.hbs

```diff
@@ -61,3 +71,16 @@ async function tryReloadImpl() {
     }
   }
 }
+
+const origMathRandom = Math.random;
+const origDateNow = Date.now;
+
+function monkeyPatchSideEffectFunctions(ctx: Context) {
+  Math.random = () => (ctx.chance as any).random(); // @types/chance is missing the definition for random
+  Date.now = () => ctx.time;
+}
```

Comment:
This is a nice start!

To improve this even further, I think I'd prefer a explicit `console.warn()` or even `throw new Error('')` that informs anyone calling these methods to consider using `ctx.time` and `ctx.chance.random()` instead.

Otherwise, we might cause unexpected behavior which could break user code without telling them why.

If you agree, implementing this would be as simple as adding a `console.warn()` before we return from our patched functions. Or we could use a Proxy for a more fancy solution (which probably isn't needed until we need to monkey patch 4-5 methods for both `Date` and `Math`). So for now, these arrow functions with a explicit warning / thrown error are good enough 🙂

Reply:
The issue is that the usage of `Math.random()` or `Date.now()` could come from an imported library that the user has no control over, and the library may not expose a way to make their functionality deterministic. Throwing an error would make it so that using the library with Hathora would be impossible.

I could consider adding a warning log, but if there's nothing the user can do about it I think it could get annoying to see it log repeatedly. Maybe a solution where it logs only once per `stateId` would work? But that seems like overkill for this PR

Reply:
Yeah, a single warning log is a good tradeoff  :+1:

Given this more thought now and it could be fine to just mention clearly in the Hathora docs that `Impl.ts` encourages deterministic methods, potentially showing a table with patched methods, and what we recommend instead.

This is ready to merge 🙂

Reply:
This is the current note in the docs around determinism: https://docs.hathora.dev/#/methods?id=determinism

Reply:
Ah, that's good enough for now! :)